Refactor USB Communication: Implement Multiple Buffers, Separate Threads, and Remove Read TimeoutsMulti buff read

dkaukov opened the pull request:
Refactored `SerialInputOutputManager`:
 
- Implemented multiple read buffers to improve USB reading efficiency and reduce latency.
- Used separate threads for reading and writing, enhancing concurrency and performance.
- Removed read timeouts, as they are not necessary for event-driven data reading.

These changes enhance the responsiveness, reliability, and robustness of USB communication.

kai-morich commented:
I wasn't aware that one can queue multiple requests on the same endpoint. Do you have numbers for the performance improvement?

I ran tests against real devices. These showed that `SerialInputOutputManager.stop()` does not return as `.requestWait()` is blocking. Please have a look into `CommonUsbSerialPort`. There `read` has some additional error checks if length 0 is returned and `close` cancels the request. Your logic should probably be moved there, with a list of `UsbRequest` instead of single one.

On recent Android versions I could observe that buffers are used round-robin. I will check with older versions too.

kai-morich commented:
crashes on Android 5.0 at
    private Supplier<UsbRequest> mRequestSupplier = UsbRequest::new;
with java.lang.NoClassDefFoundError: com.hoho.android.usbserial.util.SerialInputOutputManager$$ExternalSyntheticLambda0
	at com.hoho.android.usbserial.util.SerialInputOutputManager.<init>(SerialInputOutputManager.java:57)
	at de.kai_morich.serial_usb_terminal.SerialSocket.connect(SerialSocket.java:49)

dkaukov commented:
Hi @kai-morich,

Thanks for reviewing the changes! The refactor was driven by my work transferring audio through USB serial in [this project](https://github.com/VanceVagell/kv4p-ht), where we faced issues with USB stacks inserting zeros, especially with CP2102 "S1LABS" variants. Despite using a 16k buffer and separate thread processing, sporadic zero-filled chunks persisted. Implementing multiple small buffers fixed the issue and reduced latency, which is crucial for audio.

I’ve also addressed your suggestions: removed the Supplier for older Android compatibility and added error checks for length 0 returns.

Regarding the suggestion to move the logic to UsbSerialPort, I believe keeping SerialInputOutputManager is more appropriate. Since SerialInputOutputManager provides async functionality through callbacks, it is more logical to use the Android UsbRequest.queue(...) API there. UsbSerialPort provides blocking operations via .read() and .write(), so using UsbConnection.bulkTransfer(...) there is more appropriate. Mixing these would reduce clarity, so separating them maintains a clear distinction between async and blocking operations.

kai-morich commented:
please split the PR
- multithreading should be easy

- buffers has to be moved into UsbSerialPort, as FTDI postprocessing, error checking, ... would be missing. There could be similar method setReadBufferCount() which should be set by applications that do permanent read and should be called by default at SerialInputOutputManager.
 
If called by SerialInputOutputManager very likely has to be reconfigurable while the connection is open. Adding requests could be done immediately, for removing might be better to not re-queue them at the end of read().

I always prefer if everybody could benefit from optimizations without having to enable explicitly. Would a default re-queue at the end of read() have impact on applications only doing ocasional reads, e.g. then returning oldest instead of newest data?

dkaukov commented:
Hi @kai-morich 


Thanks for the suggestions! I have a concern about pre-filling the request queue in `UsbSerialPort`. Doing it in `.open()` could interfere with control requests, and lazily in `.read()` seems messy. 

I was thinking about adding a dedicated async API in `UsbSerialPort` for `SerialInputOutputManager` (like `.asyncReadBegin()`, `.asyncRead()`, `.asyncWrite()`), but that seems like mixing concerns without much gain. 

What would you recommend for managing the queue without causing issues with control requests or making things too complicated?

dkaukov commented:
Also, you mentioned FTDI post-processing and error checks—could you clarify which checks you're referring to and why they’re important to include in this flow?

kai-morich commented:
I have an idea for lightweight integration into `CommonUsbSerialPort`. Let me try out and send you branch for review

dkaukov commented:
Hey @kai-morich ,

I’ve pulled in the latest upstream changes to keep things up-to-date. Can you also share how to run the integration tests for this? Would be great to verify everything’s working as expected.

kai-morich commented:
[here](https://github.com/mik3y/usb-serial-for-android/wiki/Device-Tests-&-Coverage-Report) you can find instructions to run the integration tests

dkaukov commented:
@kai-morich 
It is not working very-well for me. ;( This is master branch:
<img width="602" alt="image" src="https://github.com/user-attachments/assets/d9eae102-a552-4b25-8cdb-3e9a9bc0bb18" />

The "wrongDriver()" actually crashing usb stack and "dataBits()" crashing `rfc2217_server.py`. Any advice?

kai-morich commented:
Doesn't look that bad. A few tests are always flaky. I enhanced the wiki page with missing control line loop-back connection.

What does wrongDriver crash? USB stack of Android or serial device. As some serial devices do not recover from this test I already excluded some combinations.

Where do you run rfc2217 server? I use Linux on Raspi. What is the actual error?

dkaukov commented:
Hi @kai-morich 
Every test after `wrongDriver()` was reporting "no device found". Can fix only by unplugging (an plugging back) adapter. Had to disable it. I was using Intel linux for rfc2217 server. Will post log output tomorrow. ;)

dkaukov commented:
```INFO:rfc2217.server:purge both
INFO:rfc2217.server:set baud rate: 19200
INFO:rfc2217.server:set data size: 8
INFO:rfc2217.server:set stop bits: 1
INFO:rfc2217.server:set parity: N
INFO:rfc2217.server:set baud rate: 9600
INFO:rfc2217.server:set data size: 8
INFO:rfc2217.server:set stop bits: 1
INFO:rfc2217.server:set parity: N
WARNING:rfc2217.server:ignoring Telnet command: b'\xf6'
INFO:rfc2217.server:purge both
WARNING:rfc2217.server:ignoring Telnet command: b'\xf6'
INFO:rfc2217.server:purge both
INFO:rfc2217.server:set baud rate: 19200
INFO:root:Disconnected
Traceback (most recent call last):
  File "/usr/share/doc/python3-serial/examples/./rfc2217_server.py", line 172, in <module>
    r.shortcircuit()
  File "/usr/share/doc/python3-serial/examples/./rfc2217_server.py", line 49, in shortcircuit
    self.writer()
  File "/usr/share/doc/python3-serial/examples/./rfc2217_server.py", line 79, in writer
    self.serial.write(b''.join(self.rfc2217.filter(data)))
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3/dist-packages/serial/rfc2217.py", line 1123, in filter
    self._telnet_process_subnegotiation(bytes(self.suboption))
  File "/usr/lib/python3/dist-packages/serial/rfc2217.py", line 1189, in _telnet_process_subnegotiation
    self.serial.bytesize = datasize
    ^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3/dist-packages/serial/serialutil.py", line 313, in bytesize
    self._reconfigure_port()
  File "/usr/lib/python3/dist-packages/serial/serialposix.py", line 517, in _reconfigure_port
    termios.tcsetattr(
termios.error: (22, 'Invalid argument')
```

dkaukov commented:
```Package: python3-serial
Version: 3.5-2
Priority: optional
Section: python
Source: pyserial
Origin: Ubuntu
Maintainer: Ubuntu Developers <ubuntu-devel-discuss@lists.ubuntu.com>
Original-Maintainer: Matthias Klose <doko@debian.org>
```

dkaukov commented:
Hi @kai-morich,

Looks like I've fixed the integration tests to the same level of flakiness as before. Any updates on your idea of lightweight integration?

dkaukov commented:
Hi @kai-morich,

Just wanted to check in on this PR. Is there anything else needed from my side to move it forward?

kai-morich commented:
please try https://github.com/mik3y/usb-serial-for-android/tree/readqueue branch

dkaukov commented:
Hi @kai-morich. 

I was looking into the design of `CommonUsbSerialPort` and noticed the support for read queuing with multiple buffers. However, it seems the public interface of `UsbSerialPort` doesn’t expose this functionality directly, and it's only utilized by `UsbSerialManager`. I’m curious about the reasoning behind this design choice.  

 - Was the goal to optimize throughput for everybody, even though it's abstracted away from users?  
 - Is there a plan to expose this capability in the public interface?  
 - Or is the complexity mainly for internal optimizations within `UsbSerialManager`?  
 
I’m trying to better understand the trade-offs made here, especially since this results in three working modes and quite a few conditionals in the `read()` implementation.  Any insights would be greatly appreciated!  

I will also need 2-3 days to test this implementation on a real device.

kai-morich commented:
Adding readqueue to the interface definitely makes sense, else users would have to cast or would not find this feature. 
More simplifications like defaulting bufferSize to  .getMaxPacketSize() to be considered.

dkaukov commented:
Hi @kai-morich,

I’d suggest using a more conventional approach for handling `ByteBuffer`:

```java
if(useReadQueue()) {
    buf = (ByteBuffer) response.getClientData();
    buf.flip();  // Switch to read mode
    buf.get(dest, 0, buf.limit());  // Read data into dest
    buf.clear();  // Reset for the next write
    if(mReadRequest != null) { // re-queue if connection not closed
        if (!response.queue(buf, buf.capacity())) {
            throw new IOException("Queueing USB request failed");
        }
    }
}
```
This method properly handles the buffer's position and limit, aligning with standard Java NIO practices.

kai-morich commented:
`buf.get()` does a byte wise copy, whereas `arrayCopy` is a native function that probably is faster.

dkaukov commented:
@kai-morich 
Ultimately it works. Small tweaks are here: https://github.com/mik3y/usb-serial-for-android/pull/618

kai-morich commented:
integrated, but not yet enabled by default, see commit https://github.com/mik3y/usb-serial-for-android/commit/a8b9ecc7d32ce6df749c44a2b9e8cb208ac30609
